### PR TITLE
Add NTRU + some small schema changes

### PR DESCRIPTION
**Pull request by dingens**
- Add data for scheme “NTRU”
- Make classical security level optional
- Allow nist category value of 0.

### CODE REVIEW

File: encryption/ntru/ntru-kem/bench/ref_ntruhps2048509_haswell.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+platform: >-
+  Intel Core i7-4770K (Haswell, 3.5 GHz), 32GB RAM (1333MHz), Linux 4.90-4,
+  Debian 9, GCC 6.3.0-18 (-O3)
+timings:
+  measurements:
+    gen: 12506668
+    enc|sign: 761236
+    dec|vrfy: 1940810
```
Comment (benj-zen) [inline encryption/ntru/ntru-kem/bench/ref_ntruhps2048509_haswell.yaml:8]: 19408**7**0

File: encryption/ntru/ntru-kem/param/ntruhps2048677.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+name: NTRU-HPS-2048-677
+security level:
+  quantum: 144
+  nist category: 1
+  comment: >-
+    Assuming a non-local model of computation.
+    (With a local model: 200 bits quantum security, category 3)
+sizes:
```
Comment (benj-zen) [inline encryption/ntru/ntru-kem/param/ntruhps2048677.yaml:8]: In the paper the numbers are one byte bigger (Sec. 3.2).

File: encryption/ntru/ntru-kem/param/ntruhps4096821.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+name: NTRU-HPS-4096-821
+security level:
+  quantum: 178
+  nist category: 3
+  comment: >-
+    Assuming a non-local model of computation.
+    (With a local model: 253 bits quantum security, category 5)
+sizes:
```
Comment (benj-zen) [inline encryption/ntru/ntru-kem/param/ntruhps4096821.yaml:8]: Sizes in the paper differ.

File: encryption/ntru/ntru-kem/param/ntruhrss701.yaml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+name: NTRU-HRSS-701
+security level:
+  quantum: 134
+  nist category: 1
+  comment: >-
+    Assuming a non-local model of computation.
+    (With a local model: 185 bits quantum security, category 3)
+sizes:
```
Comment (benj-zen) [inline encryption/ntru/ntru-kem/param/ntruhrss701.yaml:8]: Sizes in the paper differ.

File: encryption/ntru/ntru.yaml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  - Peter Schwabe
+  - William Whyte
+  - Zhenfei Zhang
+year:
+  candidate: 2017
+  paper: 1998
+problems:
+  - assumption: SVP
+sources:
+  - 'NIST Submission, Round 2 (https://ntru.org/f/ntru-20190330.pdf)'
```
Comment (benj-zen) [inline encryption/ntru/ntru.yaml:22]: Maybe change to a permalink

Comment (dingens) [reply encryption/ntru/ntru.yaml:24]: As the filename contains a date, I'd assume they will upload changes under a new name.

File: encryption/ntru/ntru-kem/param/ntruhps2048677.yaml (hunk 2 above)
Comment (dingens) [reply encryption/ntru/ntru-kem/param/ntruhps2048677.yaml:8]: They contradict themselves. I got the numbers from § 1.6. Supercop confirms this (https://bench.cr.yp.to/results-kem.html, 5th result when searching for `ntruhps2048677`)

File: encryption/ntru/ntru-kem/param/ntruhrss701.yaml (hunk 4 above)
Comment (dingens) [reply encryption/ntru/ntru-kem/param/ntruhrss701.yaml:8]: dito

File: encryption/ntru/ntru-kem/param/ntruhps4096821.yaml (hunk 3 above)
Comment (dingens) [reply encryption/ntru/ntru-kem/param/ntruhps4096821.yaml:8]: dito

**Comment by dingens** [comment]
Thanks. See also the comments inside the changes above.

What I'm the most uncertain about is the security levels. There are lots of numbers flying around and no simple summary. Could you independently extract numbers from the paper so that we can see whether we end up with the same result? (Rather than just checking if you find my numbers somewhere in the paper)

**Comment by benj-zen** [comment]
I'm arriving at the same numbers (Table 6, hybrid attack). They clearly understate security, since they just set o(1)=0. Also, I'm not sure if we should take the non-local model numbers. Maybe we should at least take the NIST categories from the local model, but stay with the bit costs of the non-local model (Similar to Dilithium, where the bit costs do not match the NIST levels).

**Comment by dingens** [comment]
> (Similar to Dilithium, where the bit costs do not match the NIST levels).

But there, it is because they say "our estimates are very conservative, so we put it in category X even though the numbers are a few bits too low". Here, it's two completely different models. I don't think we can mix those.

**Pull request merged**